yaaredis/pool.py

```diff
@@ -647,3 +648,235 @@ def get_node_by_slot(self, slot):
+        for _ in range(self.max_connections):
+            res.put_nowait(None)
+        return res
+
+    def reset(self):
+        """Resets the connection pool back to a clean state"""
+        self.pid = os.getpid()
+        self._available_connections = defaultdict(self._make_connection_queue)
+        self._in_use_connections = defaultdict(set)
+        self._check_lock = threading.Lock()
+        self.initialized = False
+
+    def _checkpid(self):
+        if self.pid != os.getpid():
+            with self._check_lock:
+                if self.pid == os.getpid():
+                    # another thread already did the work while we waited
+                    # on the lockself.
+                    return
+                self.disconnect()
+                self.reset()
+
+    async def get_connection(self, command_name, *keys, **options):
+        # Only pubsub command/connection should be allowed here
+        if command_name != 'pubsub':
+            raise RedisClusterException(
+                "Only 'pubsub' commands can use get_connection()")
+
+        channel = options.pop('channel', None)
+
+        if not channel:
+            return await self.get_random_connection()
+
+        slot = self.nodes.keyslot(channel)
+        node = self.get_master_node_by_slot(slot)
+
+        self._checkpid()
+
+        return await self.get_connection_by_node(node)
+
+    def make_connection(self, node):
+        """Creates a new connection"""
+
+        connection = self.connection_class(host=node['host'],
+                                           port=node['port'],
+                                           **self.connection_kwargs)
+
+        # Must store node in the connection to make it eaiser to track
+        connection.node = node
+        return connection
+
+    def release(self, connection):
+        """Releases the connection back to the pool"""
+        self._checkpid()
+        if connection.pid != self.pid:
+            return
+
+        node_name = connection.node['name']
+        self._in_use_connections[node_name].remove(connection)
+        # discard connection with unread response
+        if connection.awaiting_response:
+            connection.disconnect()
+            connection = None
+
+        try:
+            self._available_connections[node_name].put_nowait(connection)
+        except asyncio.QueueFull:
+            # perhaps the pool have been reset() ?
+            pass
+
+    def disconnect(self):
+        """Closes all connections in the pool"""
+        for queue in self._available_connections.values():
+            available_connections = []
+            while True:
+                try:
+                    connection = queue.get_nowait()
+                except asyncio.QueueEmpty:
+                    break
+                connection.disconnect()
+                available_connections.append(connection)
+
+            for connection in available_connections:
```

Comment:
Though I admit I can't think of any reason why the queue order matters, this function reverses it as a side-effect. Probably better to iterate in reversed order here to avoid it, in case users have cases where they do care about the queue implementation. `for connection in available_connections[::-1]:` should do the trick!